### PR TITLE
Update to Arrow 12.0.0, update tonic and prost

### PR DESCRIPTION
**Pull request by alamb**
This updates datafusion / ballista to use arrow 12.0.0, including upgrades to tonic / prost dependencies

Draft until arrow 12 is officially released

Also closes https://github.com/apache/arrow-datafusion/pull/2160 as it updates etcd client

### CODE REVIEW

File: .github/actions/setup-builder/action.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline .github/actions/setup-builder/action.yaml:1]: This was copied from arrow-rs , from https://github.com/apache/arrow-rs/pull/1510

**Comment by xudong963** [comment]
By the way how about upgrading etcd-client #2160😂. 
I tested it locally and changed it directly to 0.9 compiling and testing with no problems.

**Comment by alamb** [comment]
> By the way how about upgrading etcd-client https://github.com/apache/arrow-datafusion/pull/2160 in the thick😂.

Good idea -- will add

**Review by xudong963 — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
Thanks @xudong963 !

**Review by yjshen — APPROVED** [review]
Let's merge this. Thanks @alamb!

**Comment by alamb** [comment]
Thanks @yjshen

**Pull request merged**